Update for embedded-hal 1.0.0

davidstelter opened the pull request:
Maybe src/compat.rs can go away now? In any case it's building.

Close #154 (edit by @hannobraun to make sure we don't forget to close that issue when we merge this)

davidstelter committed:
Update for embedded-hal 1.0.0

Maybe src/compat.rs can go away now? In any case it's building.

hannobraun reviewed (changes requested):
Thank you for the pull request, @davidstelter! And sorry for the slow reply.

Upgrading from `embedded-hal` 1.0.0-alpha.9 to 1.0.0 is good! But in addition, this pull requests also replaces the old 0.2.7 version with 1.0.0, which seems unnecessary.

In its current form, this makes no sense, as we're still left with a `compat` module that doesn't provide any compatibility. It just translates from 1.0.0 to 1.0.0, which is redundant.

I think we have two ways to go from here:

1. Revert the removal of the `embedded-hal-stable` dependency and the changes to `compat`.
2. Remove `compat`.

Option 1 seems more reasonable to me, as that would retain compatibility with older `embedded-hal` versions at very little cost (`compat` is not a lot of code).

I don't know how prevalent pre-1.0.0 versions of `embedded-hal` still are. Maybe supporting them is no longer relevant, and I could certainly be convinced to take option 2.

What do you think?

davidstelter commented:
I found that when I merely adjusted the dependency on embedded-hal to 1.0.0 it led to build failures in src/compat.rs:

```
error[E0277]: the trait bound `<T as embedded_hal::digital::v2::OutputPin>::Error: embedded_hal::digital::Error` is not satisfied
  --> src/compat.rs:22:18
   |
22 |     type Error = T::Error;
   |                  ^^^^^^^^ the trait `embedded_hal::digital::Error` is not implemented for `<T as embedded_hal::digital::v2::OutputPin>::Error`
   |
note: required by a bound in `embedded_hal::digital::ErrorType::Error`
  --> /Users/stelterd/.cargo/registry/src/index.crates.io-6f17d22bba15001f/embedded-hal-1.0.0/src/digital.rs:61:17
   |
61 |     type Error: Error;
   |                 ^^^^^ required by this bound in `ErrorType::Error`
help: consider further restricting the associated type
   |
20 |     T::Error: fmt::Debug, <T as embedded_hal::digital::v2::OutputPin>::Error: embedded_hal::digital::Error
   |                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

It wasn't obvious to me what to do about that, which is why I just started deleting stuff. 

As for whether it's relevant to still support embedded-hal 0.2.x, the maintainers seem to think so, there's an embedded-hal-compat crate to ease interoperability: https://github.com/rust-embedded/embedded-hal/blob/master/docs/migrating-from-0.2-to-1.0.md#embedded-hal-compat

If you can see how to work around the above issues, I certainly don't object to keeping the compat code in place! Maybe it would make sense to guard it with a Cargo config macro, so it conditionally compiles only when needed? I'm not sure if a Crate feature is the appropriate way to handle that or some other mechanism. I can take a stab at it if that seems like the right course!

hannobraun commented:
Thank you for getting back to me, @davidstelter.

The `compat` module is implementing `ErrorType` from the 1.0.0 (alpha) version. Check out the difference between the alpha and final release versions of that trait:

- alpha.9: https://docs.rs/embedded-hal/1.0.0-alpha.9/embedded_hal/digital/trait.ErrorType.html
- final: https://docs.rs/embedded-hal/1.0.0/embedded_hal/digital/trait.ErrorType.html

In both versions the trait has an associated type, `Error`, but in the final version it has a new trait bound, `Error`. This trait doesn't exist in 0.2.7, thus isn't implemented for whatever T::Error is, and so we get that error, which complains about exactly that.

There's a lazy solution here, which is to do exactly what the compiler suggests in the error message, and specify that trait bound, basically replace line 19 in your version, `T::Error: fmt::Debug,`, with `T::Error: Error,` (and we'd need a `use embedded_hal::digital::Error;` up top, of course). But that would be wrong. Any old code that only implements the old `embedded-hal` traits has no reason to implement that new error trait, so us specifying that trait bound would simply mean that our implementation no longer covers any type that actually exists.

I think the solution would be to use our own error type. First, specify a new struct:

``` rust
pub struct CompatError<T> {
    pub inner: T,
}
```

That is simply a wrapper around any possible error that the old `OutputPin` implementation might return. We can implement `Error` for that pretty easily, because [all that needs to do is return an `ErrorKind`](https://docs.rs/embedded-hal/1.0.0/embedded_hal/digital/trait.Error.html), and that [`ErrorKind` only has a single variant](https://docs.rs/embedded-hal/1.0.0/embedded_hal/digital/enum.ErrorKind.html) that is always applicable:

``` rust
impl<T> Error for CompatError<T> {
    fn kind(&self) -> ErrorKind {
        ErrorKind::Other
    }
}
```

Then, specify that as the `type Error` for our `ErrorType` implementation:

``` rust
impl<T> ErrorType for Pin<T>
where
    T: OutputPin,
    T::Error: fmt::Debug,
{
    type Error = CompatError<T::Error>;
}
```

And that should be it. I tested none of that code though, so it may cause compiler errors.


> If you can see how to work around the above issues, I certainly don't object to keeping the compat code in place! Maybe it would make sense to guard it with a Cargo config macro, so it conditionally compiles only when needed? I'm not sure if a Crate feature is the appropriate way to handle that or some other mechanism. I can take a stab at it if that seems like the right course!

While it would be possible to put that code behind a feature flag, I'd rather not. It would increase the overall complexity, resulting in more stuff for the user to understand, and more configurations for the CI build to test, for example. I wouldn't agree to that unless someone can show a clear benefit. I don't think the `embedded-hal` 0.2.7 dependency is so heavy that the difference in compile times would be worth it.

It also shouldn't be necessary, because it wasn't necessary before and essentially nothing changed. We still have a new and an old version, and a module to bridge between them.

davidstelter commented:
I tried adding the following:

```
pub struct CompatError<T> {
    pub inner: T,
}

impl<T> Debug for CompatError<T> {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        Debug::fmt(&ErrorKind::Other, f)
    }
}

impl<T> Error for CompatError<T> {
    fn kind(&self) -> ErrorKind {
        ErrorKind::Other
    }
}
```

Had to add the `impl<T> Debug` part to satisfy the compiler, then faced this:

```
  --> src/compat.rs:56:9
   |
55 |     fn set_low(&mut self) -> Result<(), Self::Error> {
   |                              ----------------------- expected `Result<(), CompatError<<T as embedded_hal::digital::v2::OutputPin>::Error>>` because of return type
56 |         self.0.set_low()
   |         ^^^^^^^^^^^^^^^^ expected `Result<(), CompatError<...>>`, found `Result<(), <T as OutputPin>::Error>`
   |
   = note: expected enum `Result<_, CompatError<<T as embedded_hal::digital::v2::OutputPin>::Error>>`
              found enum `Result<_, <T as embedded_hal::digital::v2::OutputPin>::Error>`
help: consider constraining the associated type `<T as embedded_hal::digital::v2::OutputPin>::Error` to `CompatError<<T as embedded_hal::digital::v2::OutputPin>::Error>`
   |
52 |     T: StableOutputPin<Error = CompatError<<T as embedded_hal::digital::v2::OutputPin>::Error>>,
   |                       +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
```

I satisfied it by manually wrapping the error returns like so:

```
impl<T> OutputPin for Pin<T>
where
    T: StableOutputPin,
    T::Error: fmt::Debug,
{
    fn set_low(&mut self) -> Result<(), Self::Error> {
        return match self.0.set_low() {
            Ok(r) => { Ok(r) }
            Err(e) => { Err(CompatError { inner: e}) }
        }
    }

    fn set_high(&mut self) -> Result<(), Self::Error> {
        return match self.0.set_high() {
            Ok(r) => { Ok(r) }
            Err(e) => { Err(CompatError { inner: e }) }
        }
    }
}
```

And now it's compiling! Is the above acceptable?

hannobraun commented:
Thank you, @davidstelter, that's perfectly acceptable!

Feel free to just push changes as new commits, or overwrite the existing commits. That way I can directly comment on any line of code, which makes it a bit easier to give feedback.

A few notes, in case you want to make more improvements:

- Your `Debug` implementation is not ideal, as it bypasses any potential `Debug` implementation that `T` might have. This means printing the type (explicitly or as part of a panic) might omit the most important error information. I think you could replace it with two things:
  - A `#[derive(Debug)]` on `CompatError`. That will automatically implement `Debug` on `CompatError<T>`, for any `T` that already implements debug.
  - That will mean the compiler will complain again on the `impl Error`, but you can fix that by adding a `where T: Debug` to that block. That means that implementation will only be available for any `T` that implement `Debug`, but I think that's fine. It's a very pervasive trait that most types implement, and any error types _really_ should already implement it.
- You can shorten the wrapping of the errors: `self.0.set_low().map_err(|err| CompatError { inner: err })`
  - Bonus idea (not sure if that's actually better, but definitely an acceptable solution): If you turn `CompatError` into a tuple struct (`struct CompatError<T>(pub T)`), then you can shorten the error conversion further: `self.0.set_low().map_err(CompatError)`

And by the way, please let me know if it's enough already and I should shut up :grin: 

As I said above, the solution you presented is perfectly acceptable. You mentioned that you're new to Rust, and I'm just being picky for the sake of education.

davidstelter committed:
Use embedded-hal 1.0.0 and maintain 0.2.x compat.

davidstelter committed:
Add back embedded-hal 0.2.x compatibility.

davidstelter commented:
I think I've addressed all your suggestions, and I appreciate the detailed explanations and details! I was anxious about adding the type constraint for Debug, but things sure look cleaner with it. And I thought there was a nicer way to handle the error return, I've seen map_error before but had forgotten the pattern.

Thank you!

hannobraun commented on src/compat.rs:
```diff
@@ -28,10 +39,10 @@ where
 
     fn set_high(&mut self) -> Result<(), Self::Error> {
-        self.0.set_high()
+        self.0.set_low().map_err(CompatError)
```
There's a copy-paste error here:
```suggestion
        self.0.set_high().map_err(CompatError)
```

hannobraun reviewed (changes requested):
Thank you, @davidstelter! Looks very good now!

I spotted one small mistake (left a comment). I think I'll be able to accept my own suggestion after submitting this review, so probably nothing else you need to do.

hannobraun committed:
Fix `set_high`
